Docs/isd 5129: Add Tutorial Documentation (#52)
canonical/mattermost-k8s-operator: docs/tutorial.md

# Deploy the Mattermost charm for the first time

In this tutorial, we'll go through each step of the process to get a basic Mattermost deployment.

## What you'll need

You will need a working station, e.g., a laptop, with AMD64 architecture. Your working station
should have at least 4 CPU cores, 8 GB of RAM, and 50 GB of disk space.

> Tip: You can use Multipass to create a virtual machine(VM) and work in an isolated environment by running:
> ```
> multipass launch 24.04 --name mattermost-tutorial-vm --cpus 4 --memory 8G --disk 50G
> ```

### Shell into the Multipass VM
> NOTE: If you're working locally, you don't need to do this step.

To be able to work inside the Multipass VM first you need to log in with the following command:
```
multipass shell mattermost-tutorial-vm
```

> ### Install and bootstrap Juju and MicroK8s

Reviewer: ```suggestion
### Install and prepare Juju and MicroK8s
```
Nitpick, AFAIK you bootstrap a controller, not Juju

Author: Yes but I also use "bootstrap" in general as a word, too, meaning "to kickstart, prepare". But I like your suggestion!

This tutorial requires the following software to be installed on your working station
(either locally or in the Multipass VM):

- Juju 3
- MicroK8s 1.35

Use [Concierge](https://github.com/canonical/concierge) to set up Juju and MicroK8s:

```
sudo snap install --classic concierge
sudo concierge prepare -p microk8s
```

This first command installs Concierge, and the second command uses Concierge to install
and configure Juju and MicroK8s.

For this tutorial, Juju must be bootstrapped to a MicroK8s controller. Concierge should
complete this step for you, and you can verify by checking for `msg="Bootstrapped Juju" provider=microk8s`
in the terminal output and by running `juju controllers`.

If Concierge did not perform the bootstrap, run:

```
juju bootstrap microk8s tutorial-controller
```

## What you'll do

1. Deploy the Mattermost charm

> 2. Integrate with `postgresql-k8s`charm

Reviewer: ```suggestion
2. Integrate with the PostgreSQL charm
```
nitpick, to be consistent with the first step

3. Inspect the Kubernetes resources created
4. Access the Mattermost app and bootstrap
5. Clean up the environment

### Add a Juju model for the tutorial

To easily clean up the resources and separate your workload from the contents of this tutorial, set up a new Juju model named `mattermost-tutorial`:

```
juju add-model mattermost-tutorial
```

### Deploy the charms

Mattermost requires connections to PostgreSQL. For more information, see the [Charm Integrations](https://charmhub.io/mattermost-k8s/docs/reference-integrations).

Deploy the charms:
```
juju deploy postgresql-k8s --channel 14/stable --trust
juju deploy mattermost-k8s

> ```

Reviewer: Nitpick, but it feels a little weird to me that we're deploying the postgresql-k8s charm first in this Mattermost charm tutorial

Author: When I am documenting how I deploy charms for personal notes, I do like to write prerequisite charms first, and deploy the main charm as the last step. I am open to the other way around though, WDYT?

Reviewer: IMO the focus of the tutorial should be on the Mattermost charm -- we tell users they'll be deploying Mattermost for the first time, with the end goal to access a Mattermost instance. So I think we should deploy Mattermost first, and then tell users "Hey btw Mattermost requires a database, and we'll use the PostgreSQL charm for that"

I think your personal approach makes sense because, if you deploy Mattermost first, you're guaranteeing the charm to enter a `blocked` state since the prerequisite charms aren't yet deployed. But from a tutorial POV, the `blocked` state offers a learning experience for the user, especially users with little-to-no experience with Juju and charms.

Let me know if you have any thoughts about my thought process!

Author: I trust your thinking. Switched the charms around

> The Mattermost database driver requires a secure SSL/TLS connection by default. For this tutorial, we will use `self-signed-certificates` to provision a local Certificate Authority(CA) for `postgresql`:

Reviewer: ```suggestion
The Mattermost database driver requires a secure SSL/TLS connection by default. For this tutorial, we will use the `self-signed-certificates` charm to provision a local Certificate Authority(CA) for PostgreSQL:
```
nit

```
juju deploy self-signed-certificates
juju integrate postgresql-k8s self-signed-certificates:certificates

```

### Integrate with the PostgreSQL k8s charm

Integrate `postgresql-k8s` to `mattermost-k8s`:
```
juju integrate mattermost-k8s postgresql-k8s

> ```

Reviewer: Received the following error when I ran through this tutorial:
```
ERROR ambiguous relation: "mattermost-k8s postgresql-k8s" could refer to "mattermost-k8s:db postgresql-k8s:db"; "mattermost-k8s:db postgresql-k8s:db-admin"
```
Decided to use `db`, but in any event, we should clarify and update the command

self-signed-certificates           active      1  self-signed-certificates  1/stable   586  10.152.183.47   no       

Unit                         Workload  Agent  Address     Ports  Message
mattermost-k8s/0*            active    idle   10.1.1.170         
postgresql-k8s/0*            active    idle   10.1.1.180         Primary
self-signed-certificates/0*  active    idle   10.1.1.145         

Integration provider                   Requirer                       Interface          Type     Message
mattermost-k8s:secret-storage          mattermost-k8s:secret-storage  secret-storage     peer     
postgresql-k8s:database                mattermost-k8s:postgresql      postgresql_client  regular  
postgresql-k8s:database-peers          postgresql-k8s:database-peers  postgresql_peers   peer     
postgresql-k8s:restart                 postgresql-k8s:restart         rolling_op         peer     
postgresql-k8s:upgrade                 postgresql-k8s:upgrade         upgrade            peer     
self-signed-certificates:certificates  postgresql-k8s:certificates    tls-certificates   regular 

> ```

Reviewer: As a point of comparison, this is the `juju status --relations` output when I ran through the tutorial:
```
Model                Controller          Cloud/Region        Version  SLA          Timestamp
mattermost-tutorial  concierge-microk8s  microk8s/localhost  3.6.20   unsupported  13:12:03Z

App                       Version                         Status  Scale  Charm                     Channel        Rev  Address         Exposed  Message
mattermost-k8s            .../mattermost:v8.1.3-20.04...  active      1  mattermost-k8s            latest/stable   27  10.152.183.157  no       
postgresql-k8s            14.20                           active      1  postgresql-k8s            14/stable      774  10.152.183.69   no       
self-signed-certificates                                  active      1  self-signed-certificates  1/stable       586  10.152.183.45   no       

Unit                         Workload  Agent  Address       Ports     Message
mattermost-k8s/0*            active    idle   10.1.225.143  8065/TCP  
postgresql-k8s/0*            active    idle   10.1.225.141            Primary
self-signed-certificates/0*  active    idle   10.1.225.142            

Integration provider                   Requirer                       Interface         Type     Message
postgresql-k8s:database-peers          postgresql-k8s:database-peers  postgresql_peers  peer     
postgresql-k8s:db                      mattermost-k8s:db              pgsql             regular  
postgresql-k8s:restart                 postgresql-k8s:restart         rolling_op        peer     
postgresql-k8s:upgrade                 postgresql-k8s:upgrade         upgrade           peer     
self-signed-certificates:certificates  postgresql-k8s:certificates    tls-certificates  regular  
```

The deployment finishes when all the charms show `Active` states.

> Run `microk8s kubectl get pods -n mattermost-tutorial` to see the pods that are being created by the charms:

Reviewer: Ran into this error when I tried the tutorial:
```
Insufficient permissions to access MicroK8s.
You can either try again with sudo or add the user ubuntu to the 'snap_microk8s' group:

    sudo usermod -a -G snap_microk8s ubuntu
    sudo chown -R ubuntu ~/.kube

After this, reload the user groups either via a reboot or by running 'newgrp snap_microk8s'.
```

```
NAME                            READY   STATUS    RESTARTS   AGE
mattermost-k8s-0                2/2     Running   0          10m
modeloperator-64cb49db9-8cvpv   1/1     Running   0          15m
postgresql-k8s-0                2/2     Running   0          13m
self-signed-certificates-0      1/1     Running   0          13m

> ```

Reviewer: Is the term "pod" well-known, or is it specific to MicroK8s/Juju? Should we explain why postgresql and mattermost each have 2 pods?  

Also, as a nitpick, this was my output when I ran the command:
 ```
ubuntu@mattermost-tutorial-vm:~$ microk8s kubectl get pods -n mattermost-tutorial
NAME                             READY   STATUS    RESTARTS   AGE
mattermost-k8s-cc898cff9-xl6tf   1/1     Running   0          2m51s
mattermost-k8s-operator-0        1/1     Running   0          8m42s
modeloperator-6df4c76fdc-r9bvk   1/1     Running   0          8m58s
postgresql-k8s-0                 2/2     Running   0          8m34s
self-signed-certificates-0       1/1     Running   0          5m7s
```

Author: "pod" should be a well known term. Why both these applications have 2 pods is a "charm" question, as a charm with a workload would have 2 pods. I don't know if we should repeat the Juju charm documentation and explain this in this tutorial, but if we do, we'd have to do this for all our charms probably. WDYT? 

The output is for the old charm, new charm should give something similar to the tutorial!

Reviewer: Thanks for the additional context and confirming that the tutorial is correct for the new charm! 
I agree that we shouldn't duplicate Juju documentation here -- no need to add any further explanation on why some apps have 2 pods :)

> ```

Reviewer: Nitpick: I think this section should perhaps be under an H3 header, something like `### Inspect the Kubernetes resources`

> ### Access `Mattermost` for the first time 

Reviewer: ```suggestion
### Access Mattermost for the first time 
```
nit

By default, `mattermost` is exposed on the port 8080. To find the internal IP address assigned to `mattermost`, check the application address in the `juju status` output. For our example, this is `10.152.183.239`. Open a web browser, navigate to `10.152.183.239:8080` and follow the steps to set up your Mattermost server.

> > If you are using `multipass`, you need to forward the port to access the application from a web browser:

Reviewer: ```suggestion
> If you are using Multipass, you need to forward the port to access the application from a web browser:
```
nitpick

> ```

> > microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8080 -n mattermost-tutorial

Reviewer: Received this error when I ran this command:
```
ubuntu@mattermost-tutorial-vm:~$ microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8080 -n mattermost-tutorial
error: Service mattermost-k8s does not have a service port 8080
```

Reviewer: I tried updating the command to use `8065:8065` but the command hangs, so I guess that's not the solution

Reviewer: Ahhhhh okay it's supposed to "hang". I updated the command to:
```
microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8065 -n mattermost-tutorial
```

Then I used the private IP of the Multipass VM and visited the site as expected, and the terminal output updated:
```
ubuntu@mattermost-tutorial-vm:~$ microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8065 -n mattermost-tutorial
Forwarding from 0.0.0.0:8080 -> 8065
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
```

> ```

> > Then, in a seperate terminal in your host machine, you can find your VM's address by running `multipass info mattermost-tutorial-vm`. Now, you can navigate to `<your-multipass-vm-ip>:8080` to access Mattermost

Reviewer: We might want to clarify that it's the private IP address, or the first IP address listed in the output of `multipass info mattermost-tutorial-vm`

### Clean up the environment

Congratulations! You have successfully finished the Mattermost tutorial. You can now remove the
model environment that you've created using the following command:

```
juju destroy-model mattermost-tutorial --destroy-storage
```

> If you used Multipass, to remove the Multipass instance you created for this tutorial, use the following command.

Reviewer: ```suggestion
If you used Multipass, to remove the Multipass instance you created for this tutorial, use the following command on your host machine.
```
nitpick

If Concierge did not perform the bootstrap, run:

```
juju bootstrap microk8s tutorial-controller
```

## What you'll do

1. Deploy the Mattermost charm
2. Integrate with the PostgreSQL charm
3. Inspect the Kubernetes resources created
4. Access the Mattermost app and bootstrap
5. Clean up the environment

### Add a Juju model for the tutorial

To easily clean up the resources and separate your workload from the contents of this tutorial, set up a new Juju model named `mattermost-tutorial`:

```
juju add-model mattermost-tutorial
```

### Deploy the charms

Mattermost requires connections to PostgreSQL. For more information, see the [Charm Integrations](https://charmhub.io/mattermost-k8s/docs/reference-integrations).

Deploy the charms:
```
juju deploy postgresql-k8s --channel 14/stable --trust
juju deploy mattermost-k8s --channel latest/edge
```

The Mattermost database driver requires a secure SSL/TLS connection by default. For this tutorial, we will use the `self-signed-certificates` charm to provision a local Certificate Authority(CA) for PostgreSQL:
```
juju deploy self-signed-certificates
juju integrate postgresql-k8s self-signed-certificates:certificates

```

> ### Integrate with the PostgreSQL k8s charm

Reviewer: Nitpick
```suggestion
### Integrate with the PostgreSQL charm
```
Alternate suggestion:
```suggestion
### Integrate with the PostgreSQL K8s charm
```

mattermost-k8s:secret-storage          mattermost-k8s:secret-storage  secret-storage     peer     
postgresql-k8s:database                mattermost-k8s:postgresql      postgresql_client  regular  
postgresql-k8s:database-peers          postgresql-k8s:database-peers  postgresql_peers   peer     
postgresql-k8s:restart                 postgresql-k8s:restart         rolling_op         peer     
postgresql-k8s:upgrade                 postgresql-k8s:upgrade         upgrade            peer     
self-signed-certificates:certificates  postgresql-k8s:certificates    tls-certificates   regular 
```
The deployment finishes when all the charms show `Active` states.

Run `microk8s kubectl get pods -n mattermost-tutorial` to see the pods that are being created by the charms:
```
NAME                            READY   STATUS    RESTARTS   AGE
mattermost-k8s-0                2/2     Running   0          10m
modeloperator-64cb49db9-8cvpv   1/1     Running   0          15m
postgresql-k8s-0                2/2     Running   0          13m
self-signed-certificates-0      1/1     Running   0          13m
```

> > If you are using `multipass`, or installed `microk8s` from scratch, you might get "insufficient permissions" error. In that case, you can run the commands presented at the error message in your terminal, after which you can run `microk8s` commands

Reviewer: ```suggestion
> If you are using `multipass`, or installed `microk8s` from scratch, you might get an "insufficient permissions" error. In that case, run the commands presented at the error message in your terminal, after which you'll be able to run `microk8s` commands.
```
nits

> ```
> sudo usermod -a -G snap_microk8s ubuntu
> sudo chown -R ubuntu ~/.kube
> newgrp snap_microk8s
> ```

### Access Mattermost for the first time 

By default, `mattermost` is exposed on the port 8080. To find the internal IP address assigned to `mattermost`, check the application address in the `juju status` output. For our example, this is `10.152.183.239`. Open a web browser, navigate to `10.152.183.239:8080` and follow the steps to set up your Mattermost server.

> If you are using Multipass, you need to forward the port to access the application from a web browser:
> ```
> microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8080 -n mattermost-tutorial
> ```
> Then, in a separate terminal in your host machine, you can find the address of your VM by running `multipass info mattermost-tutorial-vm`. 
> This would be the private IP address and usually would be listed first. Now, you can navigate to `<your-multipass-vm-ip>:8080` to access Mattermost

### Clean up the environment

Congratulations! You have successfully finished the Mattermost tutorial. You can now remove the

> model environment that you've created using the following command:

Reviewer: Nit: Perhaps a stronger congrats message would be to describe what the user explicitly achieved by completing the tutorial, maybe something like:
```suggestion
Congratulations! You have successfully finished the Mattermost tutorial.
You have now deployed the Mattermost charm, integrated it with a database,
set up a secure SSL/TLS connection, and accessed the Mattermost instance in your browser.

You can now remove the
model environment that you've created using the following command:
```

I'm open to other ideas if you think we should focus on different achievements (or less achievements -- perhaps we only want to highlight 3 achievements instead of 4).

Author: We are "mocking" the SSL connection, but the rest, I agree. I added it.
